0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+
+Let $sparse(X)$ refer to a representation of $X$, where only the indices in the $4i$-th positions correspond to real bits, and the intermediate indices can contain auxiliary information. Connecting to the above, on input a word $X$, we can expand it in order to perform a series of boolean operations (up to 15) which results in some $sparse(X)$, encoding the same $X$. Then,
+
+$$\forall X\in\mathbb B^\ell:\quad expand(X) \equiv reset(sparse(X)) $$
+
+#### Compression
+
+Another set of functions will be defined to be used in this new design. For $i\in[0,3]$, let $mask_i(X)$ be the function that performs the AND operation $(\wedge)$ of the sparse representation of the word $2^{\ell}-1$ ("all-ones" word in binary), shifted $i$ bits to the left, with the input word $X$. More formally,
+
+$$\forall X \in (b_i)^\ell, i\in[0,3]: mask_i(X) := \left(\ expand(2^\ell - 1)<<_i 0\ \right)\ \wedge \ X$$ 
+
+meaning, on input some $sparse(X)$, all four of:
+
+$$ 
+\begin{align*}
+mask_0\left(sparse(X)\right) &:= 0 \ 0 \ 0 \ 1_{\ell-1} \ .\ .\ .\ 0 \ 0 \ 0 \ 1_{0} \ \wedge\ sparse(X) \equiv expand(X) \equiv reset(sparse(X))\\
+mask_1\left(sparse(X)\right) &:= 0 \ 0 \ 1_{\ell-1} \ 0 \ . \ . \ . \ 0 \ 0 \ 1_{0} \ 0 \ \wedge\ sparse(X) \\
+mask_2\left(sparse(X)\right) &:= 0 \ 1_{\ell-1} \ 0 \ 0 \ . \ . \ . \ 0 \ 1_{0} \ 0 \ 0 \ \wedge\ sparse(X) \\
+mask_3\left(sparse(X)\right) &:= \ 1_{\ell-1} \ 0 \ 0 \ 0 \ . \ . \ . \ 1_{0} \ 0 \ 0 \ 0 \wedge\ sparse(X) \\
+\end{align*}
+$$
+
+The effect of each $mask_i$ is to null all but the $(4n+i)$-th bits of the expanded output. Meaning, they act as selectors of every other $(4n+i)$-th bit. That implies that one can rewrite the sparse representation of any word as:
+
+$$
+\begin{align*}
+sparse(X) \iff &mask_0(sparse(X)) + mask_1(sparse(X)) \\
+& + mask_2(sparse(X)) + mask_3(sparse(X))
+\end{align*}
+$$
+
+In order to check correctness of a decomposition, we can rely on 4 lookups. The first one is a lookup to the `Reset` table using the "dense" word for the first column, and `mask_0` for the second column (because it is equivalent to the expansion). 
+The correctness of the remaining masks $(i\in[1,3])$ should also be checked. For this, the witness value $shift_i := mask_i/2^i$ will be stored in the witness in such a way that each $shift_i$ can be looked up using the "second column" of the `Reset` table (more about this later), and then the following constraint should hold:
+
+$$
+\begin{align*}
+0 = sparse(X) - (\ & reset_0(sparse(X)) + 2 \cdot shift_1(sparse(X)) \\
+& + 4 \cdot shift_2(sparse(X)) + 8 \cdot shift_3(sparse(X))\ )
+\end{align*}
+$$
+
+### Basic toolbox
+
+#### XOR
+
+The XOR of 16-bit can be computed using addition of their expansions. The 16-bit output of the XOR will be located on the $4n$ positions, whereas the intermediate positions will contain any possible carry terms.
+
+Given expanded `left` and `right` inputs, the sparse representation of $XOR_{64}$ can be constrained in quarters of 16 bits, as:
```

Comment:
To avoid ambiguity and ease the reading, I would define domains for each values, like if `left` is a 64 bits string, I would say
> Given an expanded input $\text{left} \in \mathbb{B}^{64}$

It is easier for the reader as it removes the work of checking (even if sometimes obvious) the domain of the operations.